src/psyclone/psyGen.py

```diff
@@ -6108,46 +6187,43 @@ def declare(self, name, datatype, shape=[], scope='local',
-                          moment just 'global_argument' is available for \
-                          variables passed in/out of the kernel by argument.
-        :param bool is_input: Whether the symbol represents data that exists \
-                              before the kernel is entered and that is passed \
-                              into the kernel.
-        :param bool is_output: Whether the symbol represents data that is \
-                               survives outside the kernel upon exit.
-        :raises KeyError: The provided name can not be used as key in the
+        :param interface: If not None then indicates that the symbol exists \
+                          outside the local scope and provides the details \
+                          of the sharing mechanism (e.g. routine argument).
+        :type interface: :py:class:`psyclone.psyGen.SymbolInterface` or \
+                         NoneType.
+        :raises KeyError: The provided name cannot be used as a key in the \
                           table.
         '''
         if name in self._symbols:
             raise KeyError("Symbol table already contains a symbol with"
                            " name '{0}'.".format(name))
 
-        self._symbols[name] = Symbol(name, datatype, shape, scope, is_input,
-                                     is_output)
+        self._symbols[name] = Symbol(name, datatype, shape, interface)
 
     def specify_argument_list(self, argument_name_list):
         '''
-        Keep track of the order of the arguments and provide the scope,
-        is_input and is_ouput information if it was not available on the
-        variable declaration.
+        Keep track of the order of the arguments and update the interface
+        of the associated Symbols if we weren't previously able to determine
+        their scope.
 
         :param list argument_name_list: Ordered list of the argument names.
         '''
         self._argument_list = []
         for name in argument_name_list:
             symbol = self.lookup(name)
             # Declarations without explicit intent are provisionally identified
-            # as 'local', but if they appear in the argument list the scope and
-            # input/output attributes need to be updated.
+            # as 'local', but if they appear in argument_name_list then they
+            # must have an interface added to specify that they are arguments.
             if symbol.scope == 'local':
-                symbol.scope = 'global_argument'
-                symbol.is_input = True
-                symbol.is_output = True
+                # TODO For discussion with the reviewer. The PSyIR should be
+                # language independent but obviously
+                # the line below assumes we're dealing with Fortran. Should
+                # we have some annotation somewhere that says which language
+                # we're coming from?
+                # A Fortran argument has intent(inout) by default
```

Comment:
Oh yes, that makes sense. I've altered the interface to take a SymbolInterface object. If one isn't supplied then we create one with an access of UNKNOWN. This allows the Fortran-specific front-end to specify that the default is actually READWRITE.

On a separate note, I think I've just discovered a problem with `specify_argument_list()` - it is written (and tested) so as to be called more than once. However, each time it is called it resets `self._argument_list` to [] before constructing it afresh. Therefore, if it is called more than once, e.g.:

    symbol_table.specify_argument_list(["var1"])
    symbol_table.specify_argument_list(["var2"])

then `symbol_table._argument_list` will only contain `var2` but the Symbol for `var1` remains modified to say that it is a routine argument. I think the fix for this would be to make sure that it doesn't reset `self._argument_list` and maybe call it something like `specify_arguments`/`mark_arguments` to try and make it clear that its operation is cumulative.

Reply:
You are right, when a symbol leaves the argument list  the interface should be coherent with it.

Also, the `interface` argument is just used when arguments symbols do not have an interface yet, so some are defined in the fparser2ASTProcessor and some here inside the SymbolTable.

I think I was trying to solve 2 things with this method: 1) We need to keep the order of the arguments, 2) in Fortran in some cases we just know a symbol is a Argument if it is in the argument list. Maybe it was an error to try to solve both together.

What do you think if we instead:
- Move the `for name in argument_name_list:` loop that solves (2) outside of the method into the fparser2ASTProcessor `process_declarations()` line where the method is called. (It is quite a Fortran-specific behavior anyway)
- Remove the `interface` argument as `specify_argument_list` will just be in charge of (1)
- Add 2 error conditions into `specify_argument_list`: a) All symbol names passed exist and have a Symbol.Argument interface. 2) There aren't any other symbols in the SymbolTable with interface of type Symbol.Argument.
- Add the same 2 conditions to the property that returns the argument_list. So if at some point the user adds or removes a new Symbol.Argument interface to/from a symbol, it will invalidate the `symbol_table.argument_list` until a new list order with the updated symbols is provided.